src/main_loop.rs

```diff
@@ -228,137 +232,177 @@ impl SyncState {
+    /// The decision whether a potential peer is identical to the current node
+    /// is made _only_ based on the [`InstanceId`]. This means that different
+    /// nodes that run on the same computer _can_ connect to each other.
+    ///
+    /// This method requires (immutable) access to the [`NetworkingState`],
+    /// which in turn requires the caller of this function to acquire a read
+    /// lock of the `global_state_lock`. This design is a tradeoff; the
+    /// alternative is to clone the various fields and then release the lock.
+    //
+    // A note on the design: The method uses various filters as well as a way to
+    // compare two peer candidates in a simple iterator-combinator chain to
+    // select the candidate.
+    // The design's intention is to simplify future changes: declaring a new
+    // filter as well as modifying or dropping existing filters has little to no
+    // impact on other filters. Similarly, modifying the comparison function
+    // does not require changes to the filters.
+    fn peer_candidate(&self, networking_state: &NetworkingState) -> Option<PeerCandidate> {
+        let is_self = |candidate: &PeerCandidate| candidate.id == networking_state.instance_id;
+        let same_id_is_connected = |candidate: &PeerCandidate| {
+            networking_state
+                .peer_map
+                .values()
+                .any(|peer| peer.instance_id() == candidate.id)
+        };
+        let same_socket_is_connected = |candidate: &PeerCandidate| {
+            networking_state
+                .peer_map
+                .values()
+                .filter_map(|peer| peer.listen_address())
+                .any(|address| address == candidate.address)
+        };
+        let is_bootstrap_node = |candidate: &PeerCandidate| {
+            let candidate_bootstrap_status = networking_state
+                .bootstrap_status
+                .get(&candidate.address)
+                .map(|bootstrap_info| bootstrap_info.status)
+                .unwrap_or_default();
+            candidate_bootstrap_status == BootstrapStatus::Bootstrap
         };
 
-        // Get the candidate list with the highest distance
-        let max_distance_candidates = candidates.iter().max_by_key(|pp| pp.1.distance);
+        let curr_num_peers = networking_state.peer_map.len();
+        let self_is_well_connected = self.is_well_connected(curr_num_peers);
 
-        // Pick a random candidate from the appropriate candidates
-        let mut rng = rand::rng();
-        max_distance_candidates
-            .iter()
-            .choose(&mut rng)
-            .map(|x| (x.0.to_owned(), x.1.distance))
+        self.candidates
+            .values()
+            .filter(|candidate| !is_self(candidate))
+            .filter(|candidate| !same_id_is_connected(candidate))
+            .filter(|candidate| !same_socket_is_connected(candidate))
+            .filter(|candidate| !self_is_well_connected || !is_bootstrap_node(candidate))
```

Comment:
Rewrite with De Morgan's law to `!(self_is_well_connected && is_bootstrap_node(candidate))`? I find that simpler to parse. If you disagree, leave it as it is.